mbedtls/src/x509/certificate.rs

```diff
@@ -197,6 +197,17 @@ impl LinkedCertificate {
         Ok(x509_buf_to_vec(&self.inner.subject_raw))
     }
 
+    pub fn is_self_signed(&self) -> Result<bool> {
+        if self.inner.subject_raw.p == core::ptr::null_mut() || self.inner.subject_raw.len == 0 ||
+            self.inner.issuer_raw.p == core::ptr::null_mut() || self.inner.issuer_raw.len == 0 {
+            return Err(Error::X509BadInputData)
+        }
+
+        let subject = unsafe { core::slice::from_raw_parts(self.inner.subject_raw.p, self.inner.subject_raw.len) };
+        let issuer = unsafe { core::slice::from_raw_parts(self.inner.issuer_raw.p, self.inner.issuer_raw.len) };
+        Ok(subject == issuer)
```

Comment:
This is not strictly correct because it will only find DNs which are bitwise identical. But the two DNs may actually be encoded in different ways (eg different ordering of DN components, different string types, and also IIRC the name comparisons are supposed to be be case-insensitive).

For many uses getting this wrong is not a huge problem (since in practice most CA certs are issued with identical DN bits). But for others it might cause a security problem since an attacker can construct a cert which some code will consider self-signed but this function will not consider self-signed, such as

```
peer_cert,extra_certs = parse_cert_list_sent_by_peer()
for cert in extra_certs:
   if cert.is_self_signed():
      return Error
   else:
      add_to_list_of_possible_intermediates(cert)
is_ok = validate(peer_cert)
```

If the peer sends a cert where issuer and subject DNs are identical but not bitwise identical, then this code would consider the cert not self signed, but other code (eg during certificate verification) might consider it to be self signed and treat it as a trust root.

Depending on the intended usage here I'm not sure if it is worth implementing all the required logic. Unfortunately mbedtls doesn't seem to expose a function for DN comparisons either. But at the least add a comment explaining why the logic here is incomplete and why it might cause problems re differing interpretations.

Reply:
You're right. The `is_self_signed()` function was a convenient and efficient way for me to perform the check that I need in application-level code, but - considering the concerns that you raised - it's not a function that should be part of the rust-mbedtls crate, at least not in its current form. I will remove this function and move the logic to the application.